Fix traceback when making http query to non-listening host/port and raise_error=False

terminalmage opened the pull request:
### What does this PR do?

When `salt.utils.http.query()` attempts to query a non-listening host/port and socket errors are suppressed, the response is `None`. This results in a traceback when the code attempts to log the length of the response before and after attempting to decompress it.

This PR simply returns without decompressing when the response resolves as `False`.

### What issues does this PR fix or reference?

N/A

### Previous Behavior

```
Traceback (most recent call last):
  File "/testing/salt/cli/caller.py", line 237, in call
    ret['return'] = self.minion.executors[fname](self.opts, data, func, args, kwargs)
  File "/testing/salt/executors/direct_call.py", line 12, in execute
    return func(*args, **kwargs)
  File "/testing/salt/modules/http.py", line 41, in query
    return salt.utils.http.query(url=url, opts=opts, **kwargs)
  File "/testing/salt/utils/http.py", line 618, in query
    result_text = __decompressContent(coding, result_text)
  File "/testing/salt/utils/http.py", line 108, in __decompressContent
    log.trace("Decompressing %s byte content with compression type: %s", len(pgctnt), coding)
TypeError: object of type 'NoneType' has no len()
```

### New Behavior

No traceback, and the return data is:

```python
{'body': None}
```

### Tests written?

Yes

Commit:
Fix traceback when making http query to non-listening host/port

The query result is None, which results in a traceback when we try to
report the length of the response before and after decompressing. This
fixes that by returning without attempting decompression if the response
resolves as False.

It also removes an extraneous elif which does nothing.

Commit:
Add test case

Akm0d reviewed (changes requested):
I really appreciate that you added a test case! Is returning none better than raising a more descriptive error? Maybe returning {'body': None, 'error': 'Non-listening host/port'} so that salt registers this as a failure would be appropriate.

terminalmage commented:
@Akm0d So, this only happens when `raise_error=False` to suppress the error in the `socket` module when we try to make the HTTP request. If the connection can be made, the body will be a string, instead of `None`. It seems to me like an error would be redundant.

I pushed another commit which updates the docstring for this function, and makes it more clear why `None` is returned.

Commit:
Improve documentation for http.query remote-exec func

Also, suppress full traceback when socket errors go uncaught and reraise
as CommandExecutionError.

Akm0d reviewed (approved):
Thanks for the clarification in the docs!  It makes sense to me now why you did it this way.  Looks good.

Akm0d commented on salt/modules/http.py:
```diff
@@ -35,7 +40,10 @@ def query(url, **kwargs):
+    try:
+        return salt.utils.http.query(url=url, opts=opts, **kwargs)
+    except Exception as exc:
+        raise CommandExecutionError(six.text_type(exc))
```
Pylint is failing on this line with "undefined variable six" and "undefined variable CommandExecutionError"

terminalmage replied:
Should be fixed now.

Commit:
Add missing imports
